### DISCUSSION TITLE
Cut-off dates

### DISCUSSION THREAD

**Discussion by francis-fm** [community/community#135582] [Programming Help]
### Body

I'm facing a challenging issue while working with a dataset in Python using pandas. The dataset consists of records where each row contains a specific 'sic' (identifier for promotional policies), along with 'start date' and 'end date' for when each promotion is active. My main challenge arises when handling the transitions between overlapping promotions and singular promotions.

Here's the problem: When multiple promotions overlap, I need the 'end date' for those promotions to extend through the entire last day of the overlap. However, when transitioning from a period of overlapping promotions to a period with only one active promotion, it's crucial that there's no gap or overlap in the dates. The start date of the new single promotion should exactly follow the end date of the previous promotions, without duplicating or skipping any days.

I've tried adjusting the 'end dates' to prevent overlapping or incorrect intervals by using Python's timedelta, but I'm struggling to get this right consistently, especially when the number of overlapping promotions changes. The logic needs to dynamically adjust the dates based on whether the next interval starts with a single or multiple promotions.

In summary, the problem is:  The end dates when they are of several promos policies, have problems, it is as if a day is subtracted and then, in a following row, for the one when I have only one policy, its start date is also as if a day is subtracted.

Look at tables.

My code is

````
import pandas as pd
from datetime import timedelta

# Assuming you already have the dataframe df_ap_sics
# Sorting df_ap_sics by sic, start date, and end date
df_ap_sics = df_ap_sics.sort_values(by=["sic", "start date", "end date"])

# Initialize lists for the new rows
new_rows = []

# Iterate over unique sics
for sic in df_ap_sics["sic"].unique():
    # Get all rows for this sic
    sic_rows = df_ap_sics[df_ap_sics["sic"] == sic]
    
    # Create a set of all unique dates (start and end)
    all_dates = sorted(set(sic_rows["start date"].tolist() + sic_rows["end date"].tolist()))
    
    # Adjust the dates to properly handle the intervals
    for i in range(len(all_dates) - 1):
        start = all_dates[i]
        end = all_dates[i + 1]
        
        # Correction to ensure that the end date is the day before the new start date
        # Adjust this point depending on whether the next date has more than one policy
        current_promos = set(sic_rows[(sic_rows["start date"] <= start) & (sic_rows["end date"] >= start)]["promotion policy"])
        next_promos = set(sic_rows[(sic_rows["start date"] <= end) & (sic_rows["end date"] >= end)]["promotion policy"])

        # If the current row has more than one policy, we extend the end date to include the entire day
        if len(current_promos) > 1:
            end_adjusted = end - timedelta(days=1)
        else:
            end_adjusted = end - timedelta(days=2)

        # If the next row has only one policy, adjust the start date
        if len(next_promos) == 1:
            next_start_adjusted = end
        else:
            next_start_adjusted = end - timedelta(days=1)
        
        # Get all rows that overlap with this adjusted interval
        overlapping_rows = sic_rows[(sic_rows["start date"] <= start) & (sic_rows["end date"] >= end_adjusted)]
        
        # Create a row with the unique promotions
        promos = set()
        for _, row in overlapping_rows.iterrows():
            if row["start date"] <= start and row["end date"] >= end_adjusted:
                promos.add(row["promotion policy"])
        
        # Only add a row if there are promotions
        if promos:
            new_rows.append({
                "sic": sic,
                "start date": start,
                "end date": end_adjusted,
                "promotion policy": ", ".join(promos)
            })

# Create a new dataframe with the new rows
new_df = pd.DataFrame(new_rows)

# Display the new dataframe
new_df
`````
My table is:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>

sic | start date | end date | politic_promo
-- | -- | -- | --
8189 | 15/2/2024 | 20/2/2024 | COMBO
8189 | 22/2/2024 | 27/2/2024 | COMBO
8189 | 29/2/2024 | 5/3/2024 | COMBO
8189 | 14/3/2024 | 20/3/2024 | COMBO
8189 | 21/3/2024 | 26/3/2024 | COMBO
8189 | 21/3/2024 | 30/4/2024 | HOOK
8189 | 4/4/2024 | 9/4/2024 | COMBO
8189 | 12/4/2024 | 16/4/2024 | PROMO
8189 | 12/4/2024 | 16/4/2024 | COMBO

</body>

</html>

The correct result should be:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>

sic | start date | end date | politic_promo
-- | -- | -- | --
8189 | 15/2/2024 | 20/2/2024 | COMBO
8189 | 22/2/2024 | 27/2/2024 | COMBO
8189 | 29/2/2024 | 5/3/2024 | COMBO
8189 | 14/3/2024 | 20/3/2024 | COMBO
8189 | 21/3/2024 | 26/3/2024 | COMBO, GANCHO
8189 | 27/3/2024 | 3/4/2024 | GANCHO
8189 | 4/4/2024 | 9/4/2024 | COMBO, GANCHO
8189 | 10/4/2024 | 11/4/2024 | GANCHO
8189 | 12/4/2024 | 16/4/2024 | GANCHO, PROMO, COMBO
8189 | 17/4/2024 | 30/4/2024 | GANCHO

</body>

</html>

but with the code, the incorrect result is:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>

sic | start date | end date | politic_promo
-- | -- | -- | --
8189 | 14/3/2024 | 18/3/2024 | COMBO
8189 | 20/3/2024 | 19/3/2024 | COMBO
8189 | 21/3/2024 | 25/3/2024 | COMBO, GANCHO
8189 | 26/3/2024 | 3/4/2024 | GANCHO
8189 | 4/4/2024 | 8/4/2024 | COMBO, GANCHO
8189 | 9/4/2024 | 11/4/2024 | GANCHO
8189 | 12/4/2024 | 15/4/2024 | COMBO, PROMO, GANCHO
8189 | 16/4/2024 | 29/4/2024 | GANCHO

</body>

</html>

thank u

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by everforyou** [comment]
- The end date is adjusted to ensure it covers the last day of overlapping promotions.
- The start date of new promotions is set to immediately follow the end date of the previous promotions, preventing gaps or overlaps.

```python
import pandas as pd
from datetime import timedelta

df_ap_sics = df_ap_sics.sort_values(by=["sic", "start date", "end date"])
new_rows = []

for sic in df_ap_sics["sic"].unique():
    sic_rows = df_ap_sics[df_ap_sics["sic"] == sic]
    all_dates = sorted(set(sic_rows["start date"].tolist() + sic_rows["end date"].tolist()))

    for i in range(len(all_dates) - 1):
        start = all_dates[i]
        end = all_dates[i + 1]

        current_promos = set(sic_rows[(sic_rows["start date"] <= start) & (sic_rows["end date"] >= start)]["promotion policy"])
        next_promos = set(sic_rows[(sic_rows["start date"] <= end) & (sic_rows["end date"] >= end)]["promotion policy"])

        if len(current_promos) > 1:
            end_adjusted = end - timedelta(days=1)
        else:
            end_adjusted = end

        if len(next_promos) == 1:
            next_start_adjusted = end
        else:
            next_start_adjusted = end - timedelta(days=1)

        overlapping_rows = sic_rows[(sic_rows["start date"] <= start) & (sic_rows["end date"] >= end_adjusted)]
        promos = set()

        for _, row in overlapping_rows.iterrows():
            if row["start date"] <= start and row["end date"] >= end_adjusted:
                promos.add(row["promotion policy"])

        if promos:
            new_rows.append({
                "sic": sic,
                "start date": start,
                "end date": end_adjusted,
                "promotion policy": ", ".join(promos)
            })

new_df = pd.DataFrame(new_rows)
print(new_df)
```

**Reply by francis-fm** [reply]
Thank.  The end dates when they are of several promos policies, have problems, it is as if a day is subtracted and then, in a following row, for the one when I have only one policy, its start date is also as if a day is subtracted.

The results should look like this,

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


sic | start date | end date | politic_promo
-- | -- | -- | --
8189 | 21/3/2024 | 25/3/2024 | COMBO, HOOK
8189 | 26/3/2024 | 3/4/2024 | HOOK
8189 | 4/4/2024 | 8/4/2024 | COMBO, HOOK
8189 | 9/4/2024 | 11/4/2024 | HOOK



</body>

</html>


pero con tu código y el código de arriba dan ese mismo problema 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\CMINCH~1\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


sic | start date | end date | politic_promo
-- | -- | -- | --
8189 | 21/3/2024 | 26/3/2024 | COMBO, HOOK
8189 | 27/3/2024 | 3/4/2024 | HOOK
8189 | 4/4/2024 | 9/4/2024 | COMBO, HOOK
8189 | 10/4/2024 | 11/4/2024 | HOOK



</body>

</html>
